src/datasets.c

```diff
@@ -0,0 +1,990 @@
+            SCLogNotice("rep %u", rep.value);
+
+            if (DatasetAddwRep(set, hash, 32, &rep) < 0)
+                FatalError(SC_ERR_FATAL, "die: add failed");
+            cnt++;
+        }
+    }
+
+    fclose(fp);
+    SCLogNotice("dataset: %s loaded %u records", set->name, cnt);
+    return 0;
+}
+
+static int DatasetLoadString(Dataset *set)
+{
+    if (strlen(set->load) == 0)
+        return 0;
+
+    SCLogNotice("dataset: %s loading from '%s'", set->name, set->load);
+
+    FILE *fp = fopen(set->load, "r");
+    if (fp == NULL) {
+        SCLogError(SC_ERR_DATASET, "fopen '%s' failed: %s",
+                set->load, strerror(errno));
+        return -1;
+    }
+
+    uint32_t cnt = 0;
+    char line[1024];
+    while (fgets(line, (int)sizeof(line), fp) != NULL) {
+        if (strlen(line) <= 1)
+            continue;
+
+        char *r = strchr(line, ',');
+        if (r == NULL) {
+            line[strlen(line) - 1] = '\0';
+            SCLogDebug("line: '%s'", line);
+
+            uint8_t decoded[strlen(line)];
+            uint32_t len = DecodeBase64(decoded, (const uint8_t *)line, strlen(line), 1);
+            if (len == 0)
+                FatalError(SC_ERR_FATAL, "die: bad base64 encoding");
+
+            if (DatasetAdd(set, (const uint8_t *)decoded, len) < 0)
+                FatalError(SC_ERR_FATAL, "die");
+            cnt++;
+        } else {
+            line[strlen(line) - 1] = '\0';
+            SCLogDebug("line: '%s'", line);
+
+            *r = '\0';
+
+            uint8_t decoded[strlen(line)];
+            uint32_t len = DecodeBase64(decoded, (const uint8_t *)line, strlen(line), 1);
+            if (len == 0)
+                FatalError(SC_ERR_FATAL, "die: bad base64 encoding");
+
+            r++;
+            SCLogNotice("r '%s'", r);
+
+            DataRepType rep = { .value = 0 };
+            if (ParseRepLine(r, strlen(r), &rep) < 0)
+                FatalError(SC_ERR_FATAL, "die: bad rep");
+            SCLogNotice("rep %u", rep.value);
+
+            if (DatasetAddwRep(set, (const uint8_t *)decoded, len, &rep) < 0)
+                FatalError(SC_ERR_FATAL, "die: insert failed");
+            cnt++;
+
+            SCLogNotice("line with rep %s, %s", line, r);
+        }
+    }
+
+    fclose(fp);
+    SCLogNotice("dataset: %s loaded %u records", set->name, cnt);
+    return 0;
+}
+
+extern bool g_system;
+
+enum DatasetGetPathType {
+    TYPE_STATE,
+    TYPE_LOAD,
+};
+
+static void DatasetGetPath(const char *in_path,
+        char *out_path, size_t out_size, enum DatasetGetPathType type)
+{
+    char path[PATH_MAX];
+    struct stat st;
+    int ret;
+
+    if (PathIsAbsolute(in_path)) {
+        strlcpy(path, in_path, sizeof(path));
+        strlcpy(out_path, path, out_size);
+        return;
+    }
+
+    const char *data_dir = ConfigGetDataDirectory();
+    if ((ret = stat(data_dir, &st)) != 0) {
+        SCLogNotice("data-dir '%s': %s", data_dir, strerror(errno));
```

Comment:
Probably an error here? Perhaps even failure, as something I explicitly configured is not going to be active.

Reply:
ConfigGetDataDirectory will return a default if not configured. Error will be handled later when the file is actually opened.